### DISCUSSION TITLE
[Proposal] Support for Client Side Routing

### DISCUSSION THREAD

**Discussion by blarfoon** [withastro/roadmap#223] [Proposal]
Right now Astro is the best tool in the market in my opinion. It allows the development of static sites with little to no overhead (and SSR soon), but it's missing a feature that I think is essential in today's world.

Even though big players in the scene are making quite the effort to bring MPA the same feel as SPA, we're not quite there yet and we'll not be for quite some time (https://www.youtube.com/watch?v=JCJUPJ_zDQ4).

My proposal is for Astro to support (opt-in) client-side routing and prefetching for <a> links on the page.
I don't think the JS overhead would be too much, and I think it would greatly increase the user experience when navigating websites, especially on slower networks where we could start prefetching other links as soon as the page is loaded, and while the user interacts with the page, ready up the next pages he will visit (similarly to how SvelteKit or NextJS handle it). This also allows nice between-pages transitions to be implemented for even greater UX.

**Comment by dandv** [comment]
> Right now Astro is the best tool in the market in my opinion.

I'm really interested in this - can you expand on what makes Astro the best tool, and in which market?

Gatsby is one of the mature static site generators that existed before Astro was started. It does support CSR and link prefetching, and navigation does feel instant.

What was the rationale for Astro inventing another wheel? Is this documented somewhere, i.e. the shortcomings of prior art* (including Gatsby) that Astro is trying to avoid?

\* For more on prior art, see [these slides](https://slides.com/dandv/prior-art#/5) (press :arrow_down:)

**Reply by trompx** [reply]
Just check how much js gatsby sends at initial page load vs astro + force to use graphql is meh.

**Comment by austinhallett** [comment]
I agree with this statement wholeheartedly. Astro is a fantastic tool and in truth, CSR is the only thing I have to consider when pondering the question; to astro or to not?

**Comment by matthewp** [comment]
We now have client side routing via our `<ViewTransitions />` API. See the docs for how to use it: https://docs.astro.build/en/guides/view-transitions/
